### PR TITLE
implement models with validation

### PR DESCRIPTION
**Pull request by IgorLeonenko**
This change is [Reviewable](https://reviewable.io/reviews/igorleonenko/matches/1)

### CODE REVIEW

File: app/models/match.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Match < ApplicationRecord
+
+  STATUS = ['prepare', 'in game', 'played']
```
Comment (monttda) [inline app/models/match.rb]: use the syntax of `%w(prepare played ...)` for strings arrays

File: app/models/match.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  belongs_to :home_team, class_name: 'Team'
+  belongs_to :invited_team, class_name: 'Team'
+
+  validates :game_name, presence: true, uniqueness: true
+  validates :home_team_score, presence: true, numericality: true
+  validates :invited_team_score, presence: true, numericality: true
+  validates :status, inclusion: { in: STATUS }, allow_nil: true
+  validate  :players_in_team
+
+  def players_in_team
```
Comment (monttda) [inline app/models/match.rb]: maybe a more meaningfull name that says what the method validates ? like `user_belongs_to_only_one_team_on_the_match` maybe

File: db/migrate/20160907090728_create_teams.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class CreateTeams < ActiveRecord::Migration[5.0]
+  def change
+    create_table :teams do |t|
+      t.string :name
+      t.integer :size
```
Comment (monttda) [inline db/migrate/20160907090728_create_teams.rb]: we don't need this attribute

File: db/migrate/20160907092724_create_teams_users.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class CreateTeamsUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :teams_users do |t|
+      t.belongs_to :team, index: true
+      t.belongs_to :user, index: true
+    end
```
Comment (monttda) [inline db/migrate/20160907092724_create_teams_users.rb]: we need an composite index that validates the uniqueness of the user-id , team_id (so in db can't be 2 team_users with the same user and team)

File: db/migrate/20160907093023_create_matches.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class CreateMatches < ActiveRecord::Migration[5.0]
+  def change
+    create_table :matches do |t|
+      t.string :game_name
+      t.string :result
```
Comment (monttda) [inline db/migrate/20160907093023_create_matches.rb]: Are you sure you want to score results as a string ?  in a tournament how will you easily know  who won? you will have to parse the string

File: db/migrate/20160909092420_remove_size_from_team.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class RemoveSizeFromTeam < ActiveRecord::Migration[5.0]
+  def change
```
Comment (monttda) [inline db/migrate/20160909092420_remove_size_from_team.rb]: all this migrations in which you have changed other migrations that you haven't deployed to master is better to change them in the original migration. This is because that attribute hasn't been added to the production db normally
or in the case of the dev team that noone else is working on the same branch

File: db/seeds.rb [hunk 7]
```diff
@@ -1,7 +1,31 @@
...
+TeamsUser.destroy_all
+Match.destroy_all
+
+user_names = ['jack', 'jan', 'den', 'rob']
+users      = []
+teams      = []
+
+4.times do |i|
+  users << User.create(name: "user-#{user_names[i-1]}", email: "example-#{user_names[i-1]}@mail.com",
+                       password: '123456', password_confirmation: '123456', username: "player-#{i}")
```
Comment (monttda) [inline db/seeds.rb]: maybe you want better user names that player#1 for when you will be adding them to a match for example through the interface

File: spec/factories/matches.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :match do
+    sequence(:game_name) { |n| "football-#{n}" }
```
Comment (monttda) [inline spec/factories/matches.rb]: use faker gem avoid using this #n namer (this applies to al all factories files)

**Comment by jotka87** [comment]
Reviewed 6 of 29 files at r1, 21 of 30 files at r2.
Review status: 24 of 31 files reviewed at latest revision, 18 unresolved discussions.

---

_[app/models/match.rb, line 7 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTwRjBsUFOqhEQGOSk:-KRTwRjBsUFOqhEQGOSl:by6jjih) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L7)):_

> ``` Ruby
>   belongs_to :home_team, class_name: 'Team'
>   belongs_to :invited_team, class_name: 'Team'
>   belongs_to :game
> ```

`home_team`, `invited_team` and `game` should be required

---

_[app/models/match.rb, line 12 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTwcKiTwiTiziOFXuo:-KRTwcKiTwiTiziOFXup:b-1ifxvc) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L12)):_

> ``` Ruby
>   validates :home_team_score, presence: true, numericality: true
>   validates :invited_team_score, presence: true, numericality: true
>   validates :status, inclusion: { in: STATUS }, allow_nil: true
> ```

why `allow_nil: true`? I guess that there should be always some default status, or maybe you have some good argument for that? :)

---

_[app/models/match.rb, line 19 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTxKUgc9utrSxb2Es1:-KRTxKUgc9utrSxb2Es2:b-5i19xf) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L19)):_

> ``` Ruby
> 
>     def player_can_be_only_in_one_team_on_match
>       if home_team.present? && invited_team.present? && !home_team.users.empty? && !invited_team.users.empty?
> ```

`!collection.empty?` is equal to `collection.any?` and it's more readable

---

_[app/models/match.rb, line 20 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTx0lJrX25AGR0sdij:-KRTx0lKO_lq_8OLTxJr:b-296yhz) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L20)):_

> ``` Ruby
>     def player_can_be_only_in_one_team_on_match
>       if home_team.present? && invited_team.present? && !home_team.users.empty? && !invited_team.users.empty?
>         if home_team.users == invited_team.users
> ```

What if in one team there are `[user_a, user_b]` and in the second: `[user_a, user_c]`? 

---

_[app/models/match.rb, line 27 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTyhNz0KbYZUHhu5Ll:-KRTyhNz0KbYZUHhu5Lm:b-nea4ni) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L27)):_

> ``` Ruby
> 
>     def cant_be_empty_team
>       if home_team.users.empty? || invited_team.users.empty?
> ```

this validation should go to `Team` model.

---

_[app/models/teams_user.rb, line 3 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTz2kGIlNXH6ttuLCK:-KRTz2kGIlNXH6ttuLCL:bdli75d) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/teams_user.rb#L3)):_

> ``` Ruby
> class TeamsUser < ApplicationRecord
>   belongs_to :team
>   belongs_to :user
> ```

`user` and `team` should be required.

---

_[app/models/user.rb, line 14 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTzGOMgh4dU0S8rj8j:-KRTzGOMgh4dU0S8rj8k:bt6du63) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/user.rb#L14)):_

> ``` Ruby
>   validates :username, presence: true, uniqueness: true
> 
>    mount_uploader :avatar, AvatarUploader
> ```

invalid indentation

---

_[db/migrate/20160907085345_create_users.rb, line 4 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTzj851qCdjux0RJ6l:-KRTzj851qCdjux0RJ6m:b-tyy7ar) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/db/migrate/20160907085345_create_users.rb#L4)):_

> ``` Ruby
>   def change
>     create_table :users do |t|
>       t.string :name
> ```

add `null: false` constraints to fields that are required in the model.

---

_[db/migrate/20160907092724_create_teams_users.rb, line 5 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRU-G79PJw8ogBsi3vS:-KRU-G7ASbjbN7AW5w1_:bhq22do) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/db/migrate/20160907092724_create_teams_users.rb#L5)):_

> ``` Ruby
>     create_table :teams_users do |t|
>       t.belongs_to :team, index: true
>       t.belongs_to :user, index: true
> ```

those two attributes should be required

---

_[db/migrate/20160907093023_create_matches.rb, line 5 at r1](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTMKU6_7jXyVdayLMD-r1-5:-KRU-STjSf3Kw9lUVMM4:b-tyy7ar) ([raw file](https://github.com/igorleonenko/matches/blob/c72c734930371135547b10457f3b424c703ea827/db/migrate/20160907093023_create_matches.rb#L5)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> Are you sure you want to score results as a string ?  in a tournament how will you easily know  who won? you will have to parse the string
> </details>
> add `null: false` constraints to fields that are required in the model.

---

_[db/migrate/20160912115409_create_games.rb, line 4 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRU-XdOfkMCvLYp2mnV:-KRU-XdOfkMCvLYp2mnW:b-tyy7ar) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/db/migrate/20160912115409_create_games.rb#L4)):_

> ``` Ruby
>   def change
>     create_table :games do |t|
>       t.string :name
> ```

add `null: false` constraints to fields that are required in the model.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 4 of 30 files at r2, 2 of 4 files at r3, 3 of 5 files at r4, 4 of 13 files at r5.
Review status: 19 of 31 files reviewed at latest revision, 20 unresolved discussions.

---

_[spec/models/game_spec.rb, line 4 at r3](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRY6y4o4Bcor8-eqyhz:-KRY6y4o4Bcor8-eqyi-:bays7vm) ([raw file](https://github.com/igorleonenko/matches/blob/e1600a1b8d8e008ed7d9a11ebeae55a4d5f405e5/spec/models/game_spec.rb#L4)):_

> ``` Ruby
> 
> RSpec.describe Game, type: :model do
>   describe '.validate' do
> ```

dots and hashes are used in tests description for describing class methods and instance methods accordingly, take a look here: http://betterspecs.org/#describe, so in this place it should be only "validate" or "validations".

This also applies to other test files.

---

_[spec/models/game_spec.rb, line 13 at r3](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRYAhCXrV2mHqD8Am70:-KRYAhCXrV2mHqD8Am71:b-h1cbh) ([raw file](https://github.com/igorleonenko/matches/blob/e1600a1b8d8e008ed7d9a11ebeae55a4d5f405e5/spec/models/game_spec.rb#L13)):_

> ``` Ruby
> 
>     context 'when invalid data' do
>       it 'without name' do
> ```

Test description should be constructed like a sentence, so for example "when name is not present it is invalid" - it can be represented as:

``` ruby
context "when name is not present" do
  before { game.name = nil }
  it "is invalid" do
    expect(game).not_to be_valid
  end
end
```

Basically it's the best when in `it` part there are only expects and in `before` block you can specify all the conditions for the test, so in this case assign `nil` to name. Please review other tests so it can be read in a way I just described.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 1 of 4 files at r3, 2 of 5 files at r4, 9 of 13 files at r5, 2 of 6 files at r6.
Review status: 29 of 33 files reviewed at latest revision, 16 unresolved discussions.

---

_[app/models/match.rb, line 7 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTwRjBsUFOqhEQGOSk:-KRYaTT6mTjWWnYk7fmY:b-x8lacd) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L7)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> `home_team`, `invited_team` and `game` should be required
> </details>
> you can group the same validations, like this:
> 
> ``` ruby
> validates :home_team, :invited_team, :game, presence: true
> ```

---

_[app/models/match.rb, line 12 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRTwcKiTwiTiziOFXuo:-KRYg-GWsz0RLzhgNVj7:b-ahutsg) ([raw file](https://github.com/igorleonenko/matches/blob/41de8eddd7e1ffbd603d56db91ece589befb0424/app/models/match.rb#L12)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> why `allow_nil: true`? I guess that there should be always some default status, or maybe you have some good argument for that? :)
> </details>
> Match should always have some status, right? The default one should be 'prepare'. You should make this attribute required and set the default value for it when record is initialized.

---

_[app/models/match.rb, line 21 at r5](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRYoMvR3uD6oQjWpJ55:-KRYoMvR3uD6oQjWpJ56:bki084c) ([raw file](https://github.com/igorleonenko/matches/blob/7c0bdd11ce5e234d11d223c55a53f1eaa73069a9/app/models/match.rb#L21)):_

> ``` Ruby
> 
>     def player_can_be_only_in_one_team_on_match
>       if home_team.present? && invited_team.present? && home_team.users.any? && invited_team.users.any?
> ```

you can write here something like this:

``` ruby
return if home_team.blank? || invited_team.blank?

unless home_team.user_ids & invited_team.user_ids == []
  errors.add(:base, "Same player in different teams")
end
```

it's more clear and then you're not comparing whole collections but only ids

check this out: https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals

---

_[app/models/teams_user.rb, line 1 at r5](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRY_NgBga0poJ7IOT0b:-KRY_NgBga0poJ7IOT0c:bz3ymon) ([raw file](https://github.com/igorleonenko/matches/blob/7c0bdd11ce5e234d11d223c55a53f1eaa73069a9/app/models/teams_user.rb#L1)):_

> ``` Ruby
> class TeamsUser < ApplicationRecord
> ```

Ok I just noticed that - it's incorrect name, it should be `TeamUser`. Remember to change the database table also.

---

_[spec/factories/teams.rb, line 11 at r5](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRYlSU3BrlfCAIPT-rN:-KRYlSU3BrlfCAIPT-rO:bp3zfd7) ([raw file](https://github.com/igorleonenko/matches/blob/7c0bdd11ce5e234d11d223c55a53f1eaa73069a9/spec/factories/teams.rb#L11)):_

> ``` Ruby
>       after(:build) do |team, evaluator|
>         (0...evaluator.users_count).each do |u|
>           team.users << FactoryGirl.create(:user)
> ```

you can use only `create(:user)`

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 4 of 6 files at r6.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by IgorLeonenko** [comment]
_[app/models/match.rb, line 21 at r5](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRYoMvR3uD6oQjWpJ55:-KRZ96SwNbBFBhw5jTfJ:b49w1kn) ([raw file](https://github.com/igorleonenko/matches/blob/7c0bdd11ce5e234d11d223c55a53f1eaa73069a9/app/models/match.rb#L21)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> you can write here something like this:
> 
> ``` ruby
> return if home_team.blank? || invited_team.blank?
> 
> unless home_team.user_ids & invited_team.user_ids == []
>   errors.add(:base, "Same player in different teams")
> end
> ```
> 
> it's more clear and then you're not comparing whole collections but only ids
> 
> check this out: https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals
> </details>
> wow! good thing! thank's

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

_[spec/models/match_spec.rb, line 31 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZAgHz_QdaVMYyNLdz:-KRZAgI-fT-TOo5mlVHu:bfw6w44) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L31)):_

> ``` Ruby
>     end
> 
>     context 'when without teams' do
> ```

"when no teams assigned"

---

_[spec/models/match_spec.rb, line 40 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZAssjpiZk-vLJc10J:-KRZAssjpiZk-vLJc10K:b-k8cw5g) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L40)):_

> ``` Ruby
>     end
> 
>     context 'when without game' do
> ```

"when no game assigned"

---

_[spec/models/match_spec.rb, line 43 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZB1JoWmdg60B1Pdwv:-KRZB1JoWmdg60B1Pdww:b-61lu3g) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L43)):_

> ``` Ruby
>       before { match.game = nil }
> 
>       it 'is invlaid' do
> ```

"is invalid"

---

_[spec/models/match_spec.rb, line 48 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZB3GbbOZ51nKWzoj4:-KRZB3GbbOZ51nKWzoj5:b-d647ci) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L48)):_

> ``` Ruby
>     end
> 
>     context 'when name less tahn 3 chars' do
> ```

"when name shorter than 3 chars"

---

_[spec/models/match_spec.rb, line 67 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBC9bbumdRw7MTg8W:-KRZBC9bbumdRw7MTg8X:bbaaapi) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L67)):_

> ``` Ruby
>       end
> 
>       it 'raise error' do
> ```

"raises error"

---

_[spec/models/match_spec.rb, line 73 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBEymlwiht5i9cYBA:-KRZBEymlwiht5i9cYBB:b-l93l3) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L73)):_

> ``` Ruby
>     end
> 
>     context 'when without team score' do
> ```

"when no team score assigned"

---

_[spec/models/match_spec.rb, line 84 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBKWeQVr-D-AMeJja:-KRZBKWeQVr-D-AMeJjb:b17ucp2) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L84)):_

> ``` Ruby
>     end
> 
>     context 'when contain wrong status' do
> ```

"contains"

---

_[spec/models/match_spec.rb, line 95 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBV1iZKuDdCd5ptxZ:-KRZBV1iZKuDdCd5ptx_:b-61lu3g) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/match_spec.rb#L95)):_

> ``` Ruby
>       before { match.status = nil }
> 
>       it 'without status' do
> ```

"is invalid"

---

_[spec/models/team_spec.rb, line 28 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBaprzNfwkiouUtpz:-KRZBaprzNfwkiouUtq-:b-qjzb78) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/team_spec.rb#L28)):_

> ``` Ruby
>     end
> 
>     context 'when name less than 5 chars' do
> ```

"when name shorter than 5 chars"

---

_[spec/models/team_spec.rb, line 36 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBgtDQRARON38LLZm:-KRZBgtDQRARON38LLZn:bobd39l) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/team_spec.rb#L36)):_

> ``` Ruby
>     end
> 
>     context 'when try to add same player in team' do
> ```

"when player is duplicated"

---

_[spec/models/user_spec.rb, line 22 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBq1yp70kK675Ayxq:-KRZBq1yp70kK675Ayxr:b-d647ci) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/user_spec.rb#L22)):_

> ``` Ruby
>     end
> 
>     context 'when name less than 3 chars' do
> ```

"when name shorter than 3 chars"

---

_[spec/models/user_spec.rb, line 50 at r7](https://reviewable.io:443/reviews/igorleonenko/matches/1#-KRZBuUfoKSTGZJLqhUK:-KRZBuUfoKSTGZJLqhUL:b-w0ol8b) ([raw file](https://github.com/igorleonenko/matches/blob/f2de4a55c94f02d2e63ebdc05c73d48b0f4cc595/spec/models/user_spec.rb#L50)):_

> ``` Ruby
>     end
> 
>     context 'when username less than 3 chars' do
> ```

"when username is shorter than 3 chars"

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 9 of 9 files at r9.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_

**Comment by jotka87** [comment]
Reviewed 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/1)_
